runtime/lua/vim/lsp/callbacks.lua

```diff
@@ -195,6 +195,22 @@ M['textDocument/peekDefinition'] = function(_, _, result, _)
   api.nvim_buf_add_highlight(headbuf, -1, 'Keyword', 0, -1)
 end
 
+M['textDocument/documentHighlight'] = function(_, _, result, _)
+  if not result then return end
+  local bufnr = api.nvim_get_current_buf()
+  for _, reference in ipairs(result) do
+    local start_pos = {reference["range"]["start"]["line"], reference["range"]["start"]["character"]}
+    local end_pos = {reference["range"]["end"]["line"], reference["range"]["end"]["character"]}
+    if reference["kind"] == protocol.DocumentHighlightKind.Text then
+      util.highlight_range(bufnr, util.reference_ns, "LspReferenceText", start_pos, end_pos)
+    elseif reference["kind"] == protocol.DocumentHighlightKind.Read then
+      util.highlight_range(bufnr, util.reference_ns, "LspReferenceRead", start_pos, end_pos)
+    elseif reference["kind"] == protocol.DocumentHighlightKind.Write then
+      util.highlight_range(bufnr, util.reference_ns, "LspReferenceWrite", start_pos, end_pos)
```

Comment:
Rather than expose `util.reference_ns`, this impl could go in `util`, similar to  `util.buf_diagnostics_underline`.

That also avoids the need to expose `util.highlight_range`.

Reply:
Good point